File: src/hooks/runtime/output.rs
use serde_json::{json, Value};

use crate::hooks::ir::{ClaudeEvent, HookDecision, NormalizedHookResult};

// ---- guidance-injection output formats (emitted by the `inject-guidance` hook) ----

/// Claude/Grok style: wrap the guidance body in `hookSpecificOutput.additionalContext`.
pub(crate) fn guidance_hook_specific(body: &str, event: &str) -> Value {
    json!({
        "hookSpecificOutput": {
            "hookEventName": event,
            "additionalContext": body,
        }
    })
}

/// Codex/Agy style: top-level `additionalContext` plus `continue: true`.
pub(crate) fn guidance_additional_context_continue(body: &str) -> Value {
    json!({ "additionalContext": body, "continue": true })
}

/// Cursor/OpenCode style (and the default): a bare `additional_context` field.
pub(crate) fn guidance_additional_context(body: &str) -> Value {
    json!({ "additional_context": body })
}

pub(crate) fn codex_output(result: &NormalizedHookResult) -> Value {
    let mut value = json!({
        "permissionDecision": result.decision.as_str(),
        "continue": result.decision != HookDecision::Deny,
    });
    if let Some(object) = value.as_object_mut() {
        if let Some(message) = &result.message {
            object.insert("permissionDecisionReason".to_string(), json!(message));
            object.insert("stopReason".to_string(), json!(message));
        }
        if let Some(context) = &result.additional_context {
            object.insert("additionalContext".to_string(), json!(context));
        }
        if let Some(updated_input) = &result.updated_input {
            object.insert("updatedInput".to_string(), updated_input.clone());
        }
        if let Some(updated_tool_output) = &result.updated_tool_output {
            object.insert(
                "updatedMCPToolOutput".to_string(),
                updated_tool_output.clone(),
            );
        }
        for (key, value) in &result.metadata {
            object.insert(key.clone(), value.clone());
        }
    }
    value
}

pub(crate) fn cursor_output(event: ClaudeEvent, result: &NormalizedHookResult) -> Value {
    let is_blocking = matches!(
        event,
        ClaudeEvent::PreToolUse | ClaudeEvent::PermissionRequest
    );
    let mut value = if is_blocking {
        json!({
            "permission": result.decision.as_str(),
        })
    } else {
        json!({
            "continue": result.decision != HookDecision::Deny,
        })
    };
    if let Some(object) = value.as_object_mut() {
        if let Some(message) = &result.message {
            object.insert("user_message".to_string(), json!(message));
        }
        if let Some(context) = &result.additional_context {
            object.insert("additional_context".to_string(), json!(context));
        }
        if let Some(updated_input) = &result.updated_input {
            object.insert("updated_input".to_string(), updated_input.clone());
        }
        if let Some(updated_tool_output) = &result.updated_tool_output {
            object.insert(
                "updated_mcp_tool_output".to_string(),
                updated_tool_output.clone(),
            );
        }
        for (key, value) in &result.metadata {
            object.insert(key.clone(), value.clone());
        }
    }
    value
}

pub(crate) fn claude_fallback_output(result: &NormalizedHookResult) -> Value {
    json!({
        "decision": result.decision.as_str(),
        "message": result.message,
        "additional_context": result.additional_context,
        "updated_input": result.updated_input,
        "updated_tool_output": result.updated_tool_output,
        "metadata": result.metadata,
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn guidance_format_shapes() {
        let claude = guidance_hook_specific("body text", "SessionStart");
        assert_eq!(
            claude["hookSpecificOutput"]["additionalContext"],
            "body text"
        );
        assert_eq!(
            claude["hookSpecificOutput"]["hookEventName"],
            "SessionStart"
        );

        let codex = guidance_additional_context_continue("body text");
        assert_eq!(codex["additionalContext"], "body text");
        assert_eq!(codex["continue"], true);

        let cursor = guidance_additional_context("body text");
        assert_eq!(cursor["additional_context"], "body text");
    }
}

File: internal/hooks/output.go
package hooks

import "github.com/OlegHQ/agentpack/internal/harness"

func GuidanceHookSpecific(body string, event Event) map[string]any {
	return map[string]any{"hookSpecificOutput": map[string]any{"hookEventName": string(event), "additionalContext": body}}
}

func GuidanceAdditionalContextContinue(body string) map[string]any {
	return map[string]any{"additionalContext": body, "continue": true}
}
func GuidanceAdditionalContext(body string) map[string]any {
	return map[string]any{"additional_context": body}
}

func HookOutput(target harness.Target, event Event, result Result) any {
	switch target {
	case harness.Codex, harness.Agy:
		return codexOutput(result)
	case harness.Cursor:
		return cursorOutput(event, result)
	case harness.OpenCode:
		return result
	default:
		metadata := result.Metadata
		if metadata == nil {
			metadata = map[string]any{}
		}
		return map[string]any{"decision": result.Decision, "message": nullable(result.Message), "additional_context": nullable(result.AdditionalContext), "updated_input": result.UpdatedInput, "updated_tool_output": result.UpdatedToolOutput, "metadata": metadata}
	}
}

func codexOutput(result Result) map[string]any {
	value := map[string]any{"permissionDecision": result.Decision, "continue": result.Decision != Deny}
	if result.Message != "" {
		value["permissionDecisionReason"], value["stopReason"] = result.Message, result.Message
	}
	if result.AdditionalContext != "" {
		value["additionalContext"] = result.AdditionalContext
	}
	if result.UpdatedInput != nil {
		value["updatedInput"] = result.UpdatedInput
	}
	if result.UpdatedToolOutput != nil {
		value["updatedMCPToolOutput"] = result.UpdatedToolOutput
	}
	mergeMetadata(value, result.Metadata)
	return value
}

func cursorOutput(event Event, result Result) map[string]any {
	value := make(map[string]any)
	if event == PreToolUse || event == PermissionRequest {
		value["permission"] = result.Decision
	} else {
		value["continue"] = result.Decision != Deny
	}
	if result.Message != "" {
		value["user_message"] = result.Message
	}
	if result.AdditionalContext != "" {
		value["additional_context"] = result.AdditionalContext
	}
	if result.UpdatedInput != nil {
		value["updated_input"] = result.UpdatedInput
	}
	if result.UpdatedToolOutput != nil {
		value["updated_mcp_tool_output"] = result.UpdatedToolOutput
	}
	mergeMetadata(value, result.Metadata)
	return value
}

func mergeMetadata(target, metadata map[string]any) {
	for key, value := range metadata {
		target[key] = value
	}
}
func nullable(value string) any {
	if value == "" {
		return nil
	}
	return value
}
